### ISSUE TITLE
ecs-patterns; when attachi

### ISSUE THREAD

**Issue by vjain16** [aws/aws-cdk#5286]
### Environment

  - **CDK CLI Version:** 
  - **Module Version:** 
  - **OS:** 
  - **Language:**

**Comment by pkandasamy91** [comment]
@vjain16 can you provide more information as to why you'd like to set the Target Group port differently? 

In the case of an Ec2/Fargate Load Balanced Service, the target group port is actually overridden by the container port that you define (which is why the target group port is not configurable) and most users perform TLS termination on the Load Balancer, and not the individual hosts.

**Comment by vjain16** [comment]
Yes that the point , my docker application is running on 8443 and i set the container port as 8443 but my target group is defaulting it to 80 , also if i run my docker application to http and 8080 also the target group port is still defaulting to 80 . 

I think i got the point of termination the TLS on the load balancer and pass only the http traffic to the docker application but in that case also defaulting it to port 80 is just a restriction  

But we have an existing docker application which is running on https , 8443 for this we need a way to override the protocol and port .

**Comment by pkandasamy91** [comment]
Can you provide the construct that you're initializing as well as the input properties that you're passing it?

**Comment by vjain16** [comment]
`const loadBalancedService = new ecsPatterns.ApplicationLoadBalancedEc2Service(this, name.concat('LoadBalancedSvc'), {
            memoryLimitMiB: 1024,
            cluster: mycluster,
            cpu: 512,
            certificate: cert.Certificate.fromCertificateArn(this , " my certificate" ,certificateARN ),            
            domainName : "MyDomain.com”
            domainZone : route53.HostedZone.fromHostedZoneId(this , "myZone" , “ZoneId"),
            taskImageOptions: {
                image: ecs.ContainerImage.fromEcrRepository(
                    Repository.fromRepositoryArn(this, name.concat('Repo'), ecrRepoArn),
                    dockerTag
                ),
                containerPort: containerPort,                
            },
            listenerPort : listenerPort,
            desiredCount: taskDesiredCount,
        });
               
        loadBalancedService.targetGroup.configureHealthCheck({
            port: 'traffic-port', 
            path:  healthCheckPath,
            protocol : healthCheckProtocol ,
        });
`

**Comment by rubenjgarcia** [comment]
Hi

I have the same problem

The listener of the ALB has a Target Group that points to port 80 while my container has port 6500

Is there any updates on this?

**Comment by pkandasamy91** [comment]
When using ELB and ECS together, ELB does not look at the default port specified for the target group, but rather, it looks at the container port specified. 

The default port specified on your target group is not used to route traffic from load balancer to your ecs service (the container port is used, the target group default port is ignored). The target group default port is only used if a container port is not specified (which is not possible with ECS, a container port will always be specified, even if it's the same value as the default port).

**Comment by jackm-ts** [comment]
This is also an issue for me but maybe in a slightly different way. The default props mentioned above and seen [here](https://github.com/aws/aws-cdk/blob/4e8271944468253f686bfc16e51b0d9c7bedd5bf/packages/%40aws-cdk/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts#L319-L321) and not sufficient for my application running in ecs.

The app is running container port 443. And only accepts traffic over HTTPS. The default target group protocol of HTTP is causing me to receive this error from nginx running in ECS.
```
400 The plain HTTP request was sent to HTTPS port
```
I need to be able to make the target group protocol HTTPS from `ApplicationLoadBalancedFargateServiceProps`

**Comment by sivaprasadreddy** [comment]
I am also facing the same issue. My SpringBoot app is running on port 8080 in docker container and when I used **ApplicationLoadBalancedFargateService** by default the load balancer is running on port 80 with target group of services running on port 8080. But the health checks are performed on LB on port 8080 which is failing. 

Is there a way to resolve this apart from changing the docker container port to 80?

**Comment by erothmayer** [comment]
I believe this is actually a deeper bug in the Fargate Service implementation. There are tests which appear to be asserting incorrect behavior as far as I can tell.

https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/test/fargate/test.fargate-service.ts#L1048-L1075 This test registers a service with containerPort 8000, but then asserts that the output TargetGroup in the CFN should use Port 80.  Based on commentary above, it should automatically be creating a TargetGroup with Port 8000.

That test accurately describes the broken behavior I'm seeing in my stack as well. I create an ApplicationLoadBalancedFargateService, with a containerPort of 8080. The auto-generated security groups are set up correctly -- the LB allows ingress from 0.0.0.0/0 on 443, and egress to the ECS Service SG on 8080. The ECS Service allows ingress from the LB on port 8080. But the TargetGroup incorrectly uses Port 80, and so cannot get past the security groups or reach the container.

For others finding this, the simple workaround is to use the Cfn Escape Hatch to set the TargetGroup Port manually:

```
(ecsService.targetGroup.node.defaultChild as elbv2.CfnTargetGroup).addPropertyOverride('Port', 8080);
```

**Comment by pkandasamy91** [comment]
@erothmayer happy to help!

To add a bit of clarification, the way that ECS (not specific to the CDK) works with an ALB is a bit misleading. As part of associating an ALB to an ECS service, a listener port, a target group port and a container port are configured. The listener port will be used to route external traffic to a specific target group. In the case of an ECS Service, the listener port will map to a target group, but ecs will ignore the target group's default port and instead use the container port of any containers associated to that target group. 

Example workflow:
External traffic -> calls ALB with listener port 80 -> routes traffic to target group A -> target group A ignores the default port mapping -> dynamically routes traffic to the container ports of containers mapped to the target group

That example is a bit confusing so feel free to ask more questions, but the gist of it is ECS as a service does not use the default target group port mapping. In order to have your load balancer route traffic from a different port other than 80, you'll want to modify the listener port, that routes to your container's target group.

**Comment by nragusa** [comment]
I am having a similar problem to @jackm-ts 

I understand that `containerPort` will override the port setting in the target group. However, if the protocol remains as HTTP, and the container expects an encrypted connection, it will fail. For any workload that requires end-to-end encryption, this would be a show stopper.

**Comment by pkandasamy91** [comment]
@nragusa are you looking to perform termination at the load balancer or at the container level?

**Comment by nragusa** [comment]
Ideally I'd have both. I'd like to expose the appropriate certificate to clients on the ALB, and have an encrypted connection from the ALB to the container(s).

**Comment by sep76** [comment]
Have the same problem.
that the container port override the Targetgroup port is understandable. But it still tries HTTP on a HTTPS container port. 
are there any workarounds ?

**Comment by nickorka** [comment]
I've specified TCP protocol for Fargate service, but I'm getting another error now:
`Value 'tcp' at 'protocol' failed to satisfy constraint: Member must satisfy enum value set: [UDP, TCP, TCP_UDP, HTTP, HTTPS, TLS] (Service: ElasticLoadBalancingV2, Status Code: 400`

Here is my code:

```python
vpc = ec2.Vpc.from_lookup(self, 'VPC', is_default=False, vpc_name='Blah-vpc')

        cluster = ecs.Cluster(
            self, 'LuigiServerCluster', vpc=vpc,
            enable_fargate_capacity_providers=True)

        image = ecs.ContainerImage.from_asset('../container/')

        fargate_service = ecs_patterns.ApplicationLoadBalancedFargateService(
            self, "LugiServerService",
            cluster=cluster,  # Required
            cpu=512,  # Default is 256
            desired_count=1,  # Default is 1
            task_image_options=ecs_patterns.ApplicationLoadBalancedTaskImageOptions(
                image=image, container_port=8082
            ),
            memory_limit_mib=2048,  # Default is 512
            public_load_balancer=False,  # Default is False
            protocol=ecs.Protocol.TCP,
            listener_port=8082
        )
```
What is wrong?

**Comment by serge-komogorov-idealo** [comment]
@[nickorka](https://github.com/nickorka), 
protocol: ApplicationProtocol.HTTPS, where ApplicationProtocol comes from @aws-cdk/aws-elasticloadbalancingv2
